[GPU Process] [FormControls] Add a ControlPart for TextField

shallawa opened the pull request:
#### 8209e22dca2cc1df925cf2aae58a982a66a93dde
<pre>
[GPU Process] [FormControls] Add a ControlPart for TextField
<a href="https://bugs.webkit.org/show_bug.cgi?id=249573">https://bugs.webkit.org/show_bug.cgi?id=249573</a>
rdar://103508637

Reviewed by Simon Fraser and Aditya Keerthi.

NSTextFieldCell is used draw a border around the text field. CoreUI is used to
draw a list button if the text field has a data list.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/mac/NSTextFieldCellSPI.h: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlPart.h:
* Source/WebCore/platform/graphics/controls/ControlStyle.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/controls/TextFieldPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
(WebCore::TextFieldPart::create):
(WebCore::TextFieldPart::TextFieldPart):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::levelIndicatorCell const):
(WebCore::ControlFactoryMac::textFieldCell const):
(WebCore::ControlFactoryMac::createPlatformTextField):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::userPrefersContrast):
(WebCore::drawCellOrFocusRing):
(WebCore::ControlMac::drawCell):
(WebCore::ControlMac::drawListButton):
(WebCore::ControlMac::draw): Deleted.
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
(WebCore::MeterMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.mm: Added.
(WebCore::TextFieldMac::TextFieldMac):
(WebCore::TextFieldMac::shouldPaintCustomTextField):
(WebCore::TextFieldMac::draw):
* Source/WebCore/platform/graphics/mac/controls/WebControlView.h:
* Source/WebCore/platform/graphics/mac/controls/WebControlView.mm:
(-[WebControlTextFieldCell _adjustedCoreUIDrawOptionsForDrawingBordersOnly:]):
(-[WebControlTextFieldCell _coreUIDrawOptionsWithFrame:inView:includeFocus:]):
(-[WebControlTextFieldCell _coreUIDrawOptionsWithFrame:inView:includeFocus:maskOnly:]):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::ensureControlPartForRenderer):
(WebCore::RenderBox::ensureControlPartForBorderOnly):
(WebCore::RenderBox::paintBoxDecorations):
(WebCore::RenderBox::ensureControlPart): Deleted.
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::hasListButton const):
(WebCore::RenderTheme::hasListButtonPressed const):
(WebCore::RenderTheme::createControlPartForRenderer const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::canCreateControlPartForBorderOnly const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForBorderOnly const):
(WebCore::RenderThemeMac::meterSizeForBounds const):
(WebCore::RenderThemeMac::shouldPaintCustomTextField const): Deleted.
(WebCore::RenderThemeMac::paintTextField): Deleted.
(WebCore::RenderThemeMac::textField const): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/258125@main">https://commits.webkit.org/258125@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b89ef51cb89e6392a17e19e798c8464c2839eb4e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100974 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34025 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110278 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170538 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11059 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/999 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93381 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108114 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106757 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8368 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91625 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34971 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90276 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23025 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3801 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24542 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3829 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/944 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9940 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44046 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5578 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5597 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
```diff
@@ -66,11 +67,31 @@
+NSTextFieldCell* ControlFactoryMac::textFieldCell() const
+{
+    if (!m_textFieldCell) {
+        m_textFieldCell = adoptNS([[WebControlTextFieldCell alloc] initTextCell:@""]);
```
Wrap in `BEGIN_BLOCK_OBJC_EXCEPTIONS`/`END_BLOCK_OBJC_EXCEPTIONS`

shallawa replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -43,6 +47,11 @@
 
+bool ControlMac::userPrefersContrast()
+{
+    return [[NSWorkspace sharedWorkspace] accessibilityDisplayShouldIncreaseContrast];
```
Which process does this run in? If it's in the GPU Process, we should feed it in from the web process.

shallawa replied:
I will address this issue in a separate patch.

smfr commented on Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h:
```diff
@@ -0,0 +1,48 @@
+
+namespace WebCore {
+
+class TextFieldMac :  public ControlMac {
```
`final`

smfr commented on Source/WebCore/platform/graphics/mac/controls/WebControlView.h:
```diff
@@ -37,4 +37,12 @@
 @property (class) NSRect clipBounds;
 @end
 
+@interface NSTextFieldCell ()
```
Rather than copy this here, move it to an SPI header.

shallawa replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1165,6 +1182,29 @@ bool RenderTheme::isDefault(const RenderObject& o) const
 }
 
+#if ENABLE(DATALIST_ELEMENT)
+bool RenderTheme::hasListButton(const RenderObject& o) const
```
`o` -> `renderer`

shallawa replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1165,6 +1182,29 @@ bool RenderTheme::isDefault(const RenderObject& o) const
+    return input.list();
+}
+
+bool RenderTheme::hasListButtonPressed(const RenderObject& o) const
```
Ditto

shallawa replied:
Fixed.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h:
```diff
@@ -0,0 +1,48 @@
+
+namespace WebCore {
+
+class TextFieldMac :  public ControlMac {
```
Nit: There's an extra space after the `:`.

shallawa replied:
Space was removed.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h:
```diff
@@ -0,0 +1,48 @@
+
+class TextFieldMac :  public ControlMac {
+public:
+    TextFieldMac(TextFieldPart& owningPart, ControlFactoryMac&, NSTextFieldCell*);
```
Nit: `NSTextFieldCell *` since it's an Objective-C type.

shallawa replied:
Fixed.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -626,8 +640,11 @@ bool RenderTheme::paint(const RenderBox& box, ControlStates& controlStates, cons
 
-    if (UNLIKELY(!canPaint(paintInfo, box.settings(), type)))
+    if (UNLIKELY(!canPaint(paintInfo, box.settings(), type))) {
+        if (type == ControlPartType::MenulistButton || type == ControlPartType::TextField || type == ControlPartType::TextArea || type == ControlPartType::Listbox)
```
Why is this change not being made in the `canPaint` method? Also, why are unrelated types being added in this patch?

shallawa replied:
Fixed. Check was moved to canPaint().

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1165,6 +1179,29 @@ bool RenderTheme::isDefault(const RenderObject& o) const
+#if ENABLE(DATALIST_ELEMENT)
+bool RenderTheme::hasListButton(const RenderObject& renderer) const
+{
+    if (!is<HTMLInputElement>(renderer.generatingNode()))
+        return false;
+
+    const auto& input = downcast<HTMLInputElement>(*(renderer.generatingNode()));
```
`dynamicDowncast`

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1165,6 +1179,29 @@ bool RenderTheme::isDefault(const RenderObject& o) const
+
+bool RenderTheme::hasListButtonPressed(const RenderObject& renderer) const
+{
+    const auto* input = downcast<HTMLInputElement>(renderer.generatingNode());
+    if (!input)
+        return false;
```
`dynamicDowncast`

shallawa committed:
[GPU Process] [FormControls] Add a ControlPart for TextField
https://bugs.webkit.org/show_bug.cgi?id=249573
rdar://103508637

Reviewed by Simon Fraser and Aditya Keerthi.

NSTextFieldCell is used draw a border around the text field. CoreUI is used to
draw a list button if the text field has a data list.

* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/mac/NSTextFieldCellSPI.h: Added.
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlPart.h:
* Source/WebCore/platform/graphics/controls/ControlStyle.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/controls/TextFieldPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
(WebCore::TextFieldPart::create):
(WebCore::TextFieldPart::TextFieldPart):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::levelIndicatorCell const):
(WebCore::ControlFactoryMac::textFieldCell const):
(WebCore::ControlFactoryMac::createPlatformTextField):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::userPrefersContrast):
(WebCore::drawCellOrFocusRing):
(WebCore::ControlMac::drawCell):
(WebCore::ControlMac::drawListButton):
(WebCore::ControlMac::draw): Deleted.
* Source/WebCore/platform/graphics/mac/controls/MeterMac.mm:
(WebCore::MeterMac::draw):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.mm: Added.
(WebCore::TextFieldMac::TextFieldMac):
(WebCore::TextFieldMac::shouldPaintCustomTextField):
(WebCore::TextFieldMac::draw):
* Source/WebCore/platform/graphics/mac/controls/WebControlView.h:
* Source/WebCore/platform/graphics/mac/controls/WebControlView.mm:
(-[WebControlTextFieldCell _adjustedCoreUIDrawOptionsForDrawingBordersOnly:]):
(-[WebControlTextFieldCell _coreUIDrawOptionsWithFrame:inView:includeFocus:]):
(-[WebControlTextFieldCell _coreUIDrawOptionsWithFrame:inView:includeFocus:maskOnly:]):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::ensureControlPartForRenderer):
(WebCore::RenderBox::ensureControlPartForBorderOnly):
(WebCore::RenderBox::paintBoxDecorations):
(WebCore::RenderBox::ensureControlPart): Deleted.
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::hasListButton const):
(WebCore::RenderTheme::hasListButtonPressed const):
(WebCore::RenderTheme::createControlPartForRenderer const): Deleted.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::canCreateControlPartForBorderOnly const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForBorderOnly const):
(WebCore::RenderThemeMac::meterSizeForBounds const):
(WebCore::RenderThemeMac::shouldPaintCustomTextField const): Deleted.
(WebCore::RenderThemeMac::paintTextField): Deleted.
(WebCore::RenderThemeMac::textField const): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/258125@main
